proposals/ImplicitCopyWarnings.md

```diff
@@ -0,0 +1,144 @@
+can only happen on the CPU. Then this training loop blocks on data transfer and
+CPU computation every iteration:
+
+```swift
+public func train(inputs: Tensor<Float>, outputs: Tensor<Float>, initialWeights: Tensor<Float>) -> Tensor<Float> {
+  var weights = initialWeights
+  for step in 0...1000 {
+    let predictions = cpuOnlyComputation(inputs • weights)
+    let errors = predictions - outputs
+    let dweights = (errors * inputs).sum(alongAxes: 0).transposed()
+    weights -= 0.01 * dweights
+
+    if (step % 100 == 0) {
+      print("Current weights: \(weights)")
+    }
+  }
+  return weights
+}
+```
+
+We should emit a warning in the above code so that the user is aware that their
+training loop is blocking on data transfer and CPU computation.
+
+The round-trip-rule achieves exactly what we want in these examples! So does the
+heuristic of warning for all transfers from the host to the accelerator.
+
+## False-positive / False-negative tradeoff
+
+The round-trip-rule does not catch all slowness related to implicit copies. For
+example, a program that frequently dumps large pieces of data from the GPU to
+the CPU might soak up GPU memory with data waiting to be copied, and the
+round-trip-rule will not catch that.
+
+However, the round-trip-rule does capture what I currently believe will be the
+main source of performance unpredictability (accidentally writing code that
+moves a piece of a computation onto a different device), so I propose that it's
+a good initial balance between false positives and false negatives.
+
+After we gain more experience with Swift models, we can revisit implicit copy
+warnings and see whether the balance still appears to be good.
+
+## Issues with the heuristic
+
+The heuristic (warn for all transfers from the host to the accelerator) can
+produce false positive warnings for harmless transfers from the host to the
+accelerator. One common situation where this happens is when the host calculates
+some simple control flow conditions and sends them to the accelerator. For
+example:
+
+```swift
+public func example(steps: Int) -> Tensor<Float> {
+  var result: Tensor<Float> = Tensor(zeros: [10, 10])
+  vat step: Int = 0
+  while step < steps {
+    if step % 2 == 0 {
+      result += 1
+    } else {
+      result += 2
+    }
+    step += 1
+  }
+  return result
+}
+```
+
+If `step < steps` and `step % 2 == 0` get evaluated on the host and the boolean
+results get copied over to the accelerator, then the heuristic will warn about
+implicit copies. But these implicit copies are harmless because the host can
+quickly run through a bunch of iterations and queue up a bunch of booleans for
+the accelerator to consume.
+
+Without implementing the true round-trip-rule, we can denoise the warnings in
+that situation by suppressing warnings for scalar transfers. But the
+round-trip-rule suppresses those warnings in a cleaner and more reliable way, so
+I propose that we eventually do implement the round-trip-rule.
```

Comment:
another possible future direction that you mentioned in our earlier discussion is: 
Move sends/recvs diagnostics to a separate tool, possibly using runtime info (e.g.  profile data).

Do you want to also briefly cover it in this doc, to establish the "space" of our design exploration?

Reply:
Will add!